File: mcp-client-bridge/src/health-check-manager.ts
/**
 * Health Check Manager — Periodic ping to verify Orchestrator connectivity.
 * Triggers reconnection when ping fails.
 */

import { BridgeConfig } from './bridge-config.js';
import { HttpStreamableClient } from './http-streamable-client.js';
import { ReconnectionManager, BridgeState } from './reconnection-manager.js';

export interface HealthCheckConfig {
  pingIntervalMs: number;       // default 30000 (0 = disabled)
  pingTimeoutMs: number;        // default 5000
}

export class HealthCheckManager {
  private timer: ReturnType<typeof setTimeout> | null = null;
  private pingId = 0;
  private consecutiveFailures = 0;

  constructor(
    private readonly config: HealthCheckConfig,
    private readonly httpClient: HttpStreamableClient,
    private readonly reconnectionManager: ReconnectionManager,
  ) {}

  start(): void {
    if (this.config.pingIntervalMs <= 0) {
      console.error('[mcp-bridge] Health check disabled (interval=0)');
      return;
    }
    this.consecutiveFailures = 0;
    this.schedulePing();
    console.error(`[mcp-bridge] Health check started (interval=${this.config.pingIntervalMs}ms)`);
  }

  stop(): void {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }

  private schedulePing(): void {
    this.timer = setTimeout(async () => {
      await this.executePing();
    }, this.config.pingIntervalMs);
  }

  private async executePing(): Promise<void> {
    const ok = await this.sendPing();
    if (ok) {
      this.onPingSuccess();
    } else {
      this.onPingFailure();
    }
  }

  private async sendPing(): Promise<boolean> {
    try {
      const response = await this.httpClient.sendRequest('ping');
      // Any JSON-RPC response (even error) means server is alive
      return response !== null && response !== undefined;
    } catch {
      return false;
    }
  }

  private onPingSuccess(): void {
    if (this.consecutiveFailures > 0) {
      console.error('[mcp-bridge] Ping OK — connection restored');
    }
    this.consecutiveFailures = 0;
    this.pingId++;
    if (this.pingId % 10 === 1) {
      console.error(`[mcp-bridge] Health check OK (ping #${this.pingId})`);
    }
    this.schedulePing();
  }

  private onPingFailure(): void {
    this.consecutiveFailures++;
    console.error(
      `[mcp-bridge] Ping failed (${this.consecutiveFailures} consecutive failure${this.consecutiveFailures > 1 ? 's' : ''})`,
    );

    if (this.consecutiveFailures >= 1 && this.reconnectionManager.state !== BridgeState.RECONNECTING) {
      this.triggerReconnect();
    } else {
      // Already reconnecting or below threshold — just schedule next ping
      this.schedulePing();
    }
  }

  private triggerReconnect(): void {
    this.stop();
    console.error('[mcp-bridge] State: CONNECTED → DISCONNECTED (ping timeout)');
    this.reconnectionManager.state = BridgeState.DISCONNECTED;
    this.httpClient.resetSession();
    this.reconnectionManager.reconnectLoop().then(() => {
      if (this.reconnectionManager.state === BridgeState.CONNECTED) {
        console.error('[mcp-bridge] Reconnected — resuming health check');
        this.consecutiveFailures = 0;
      }
      // Always restart health check regardless of reconnect result
      this.schedulePing();
    }).catch(() => {
      // Reconnect loop failed — still restart health check to keep reporting
      console.error('[mcp-bridge] Reconnect failed — health check will continue reporting');
      this.schedulePing();
    });
  }
}

File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/HealthCheckManager.kt
package com.orchestrator.mcp.bridge

import kotlinx.coroutines.*
import org.slf4j.LoggerFactory
import java.util.concurrent.atomic.AtomicInteger

/**
 * Health Check Manager — Periodic ping to verify Orchestrator connectivity.
 * Triggers reconnection when ping fails.
 */
class HealthCheckManager(
    private val config: HealthCheckConfig,
    private val httpClient: HttpStreamableClient,
    private val reconnectionManager: ReconnectionManager
) {
    private val logger = LoggerFactory.getLogger(HealthCheckManager::class.java)
    private val consecutiveFailures = AtomicInteger(0)
    private val pingId = AtomicInteger(0)
    private var healthCheckJob: Job? = null

    fun start(scope: CoroutineScope) {
        if (config.pingIntervalMs <= 0) {
            logger.info("Health check disabled (interval=0)")
            return
        }
        consecutiveFailures.set(0)
        healthCheckJob = scope.launch { pingLoop() }
        logger.info("Health check started (interval=${config.pingIntervalMs}ms)")
    }

    fun stop() {
        healthCheckJob?.cancel()
        healthCheckJob = null
    }

    private suspend fun pingLoop() {
        while (currentCoroutineContext().isActive) {
            delay(config.pingIntervalMs)
            val ok = sendPing()
            if (ok) {
                onPingSuccess()
            } else {
                onPingFailure()
            }
        }
    }

    private suspend fun sendPing(): Boolean {
        return try {
            httpClient.sendRequest("ping", null)
            true
        } catch (_: Exception) {
            false
        }
    }

    private fun onPingSuccess() {
        val prev = consecutiveFailures.getAndSet(0)
        if (prev > 0) {
            logger.info("Ping OK — connection restored")
        }
    }

    private suspend fun onPingFailure() {
        val failures = consecutiveFailures.incrementAndGet()
        logger.warn("Ping failed ($failures consecutive)")

        if (failures >= config.failureThreshold) {
            triggerReconnect()
        }
    }

    private suspend fun triggerReconnect() {
        stop()
        logger.info("State: CONNECTED → DISCONNECTED (ping timeout)")
        reconnectionManager.state = BridgeState.DISCONNECTED
        httpClient.resetSession()
        reconnectionManager.reconnectLoop()
        if (reconnectionManager.state == BridgeState.CONNECTED) {
            val scope = CoroutineScope(currentCoroutineContext())
            start(scope)
        }
    }
}
